### PR TITLE
feat(swarm): add `peer_id` to `ListenFailure`

### PR DESCRIPTION
**Pull request by nathanielc**
## Description

It's possible in certain failure modes to know the peer_id of a failed incoming connection. For example when an inbound connection has been negotiated/upgraded but then rejected locally for a connection limit or similar reason. In these cases it makes sense to communicate the peer_id to behaviours in case they have created any internal state about the peer.

Related https://github.com/libp2p/rust-libp2p/pull/4777

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/request-response/src/lib.rs [hunk 1]
```diff
@@ -677,7 +677,24 @@ where
         }
     }
 
-    fn on_dial_failure(&mut self, DialFailure { peer_id, .. }: DialFailure) {
+    // Removes the connection for a peer if it exists.
+    fn remove_connection_for_peer(&mut self, peer: PeerId, connection_id: ConnectionId) {
```
Comment (thomaseizinger) [inline protocols/request-response/src/lib.rs]: This is unrelated to adding the field, right?

**Review by thomaseizinger — COMMENTED** [review]
Thanks! This is a breaking change so'll have to hold off on it for a while. I hope that is okay. https://github.com/libp2p/rust-libp2p/pull/4777 is not blocked on this as far as I understand because `ConnectionId`s are unique across peers so we should be able to clean up state properly!

File: protocols/request-response/src/lib.rs (hunk 1 above)
Comment (nathanielc) [reply protocols/request-response/src/lib.rs]: Oops I thought I split the commits, must have messed something up. I'll get this removed from this PR and update #4777  to not depend on these changes.

**Comment by nathanielc** [comment]
> This is a breaking change so'll have to hold off on it for a while. I hope that is okay.

Yep all good,no rush on this feature from us.

**Review by jxs — APPROVED** [review]
Thanks!
